src/blas/backends/cublas/cublas_scope_handle.cpp

```diff
@@ -17,126 +17,28 @@
-#include <CL/sycl/detail/common.hpp>
-#endif
 
 namespace oneapi {
 namespace mkl {
 namespace blas {
 namespace cublas {
 
-/**
- * Inserts a new element in the map if its key is unique. This new element
- * is constructed in place using args as the arguments for the construction
- * of a value_type (which is an object of a pair type). The insertion only
- * takes place if no other element in the container has a key equivalent to
- * the one being emplaced (keys in a map container are unique).
- */
-#ifdef ONEMKL_PI_INTERFACE_REMOVED
-thread_local cublas_handle<ur_context_handle_t> CublasScopedContextHandler::handle_helper =
-    cublas_handle<ur_context_handle_t>{};
-#else
-thread_local cublas_handle<pi_context> CublasScopedContextHandler::handle_helper =
-    cublas_handle<pi_context>{};
-#endif
-
-CublasScopedContextHandler::CublasScopedContextHandler(sycl::queue queue, sycl::interop_handle& ih)
+CublasScopedContextHandler::CublasScopedContextHandler(sycl::interop_handle& ih)
         : ih(ih),
-          needToRecover_(false) {
-    placedContext_ = new sycl::context(queue.get_context());
-    auto cudaDevice = ih.get_native_device<sycl::backend::ext_oneapi_cuda>();
-    CUresult err;
-    CUcontext desired;
-    CUDA_ERROR_FUNC(cuCtxGetCurrent, err, &original_);
-    CUDA_ERROR_FUNC(cuDevicePrimaryCtxRetain, err, &desired, cudaDevice);
-    if (original_ != desired) {
-        // Sets the desired context as the active one for the thread
-        CUDA_ERROR_FUNC(cuCtxSetCurrent, err, desired);
-        // No context is installed and the suggested context is primary
-        // This is the most common case. We can activate the context in the
-        // thread and leave it there until all the PI context referring to the
-        // same underlying CUDA primary context are destroyed. This emulates
-        // the behaviour of the CUDA runtime api, and avoids costly context
-        // switches. No action is required on this side of the if.
-        needToRecover_ = !(original_ == nullptr);
-    }
-}
-
-CublasScopedContextHandler::~CublasScopedContextHandler() noexcept(false) {
-    if (needToRecover_) {
-        CUresult err;
-        CUDA_ERROR_FUNC(cuCtxSetCurrent, err, original_);
-    }
-    delete placedContext_;
-}
-
-void ContextCallback(void* userData) {
-    auto* ptr = static_cast<std::atomic<cublasHandle_t>*>(userData);
-    if (!ptr) {
-        return;
-    }
-    auto handle = ptr->exchange(nullptr);
-    if (handle != nullptr) {
-        cublasStatus_t err1;
-        CUBLAS_ERROR_FUNC(cublasDestroy, err1, handle);
-        handle = nullptr;
-    }
-    else {
-        // if the handle is nullptr it means the handle was already destroyed by
-        // the cublas_handle destructor and we're free to delete the atomic
-        // object.
-        delete ptr;
-    }
+          nativeDevice(ih.get_native_device<sycl::backend::ext_oneapi_cuda>()) {
+    cublasStatus_t err;
+    CUBLAS_ERROR_FUNC(cublasCreate, err, &cublasHandle);
```

Comment:
This is a big change, the cublas handle used to be cached in the global map so it is created only once per device and host thread. Now cublas handles will be created in every host_task (and never destroyed, see my previous comment).
I think we should keep the ability to cache the cublas handles.

Reply:
What do you mean that it will be created in every host_task? Does that mean that in a typical application which uses this, we would call several host_tasks in a single thread where each of those would instantiate `CublasScopedContextHandler`?

Reply:
> What do you mean that it will be created in every host_task?

What I mean is that every compute function will launch a `host_task` which will call the constructor of `CublasScopedContextHandler` [here](https://github.com/oneapi-src/oneMKL/blob/develop/src/blas/backends/cublas/cublas_task.hpp#L75).

A typical application will have a main thread which will submit multiple `host_task`. From there it is hard to predict how many threads will be used, it depends on the SYCL runtime and on the graph dependencies. I believe DPC++ will try to re-use the same thread for 2 consecutive `host_task` with a linear dependencies. I'm not 100% sure what happens to the lifetime of the different objects when the thread is re-used.
I believe that with the version in upstream the application may end up having multiple cublas handle for each thread in the DPC++ thread pool (which is maybe not ideal but seems safer) so the number of cublas handle would end up bounded by the number of threads used by DPC++.
I believe that with your suggested solution every `host_task` submitted will simply create a new cublas handle i.e. the more functions from oneMKL BLAS are called the more cublas handle will be created.

Reply:
Right, I think I see what potentially could be a problem here and why the destruction of the cublas handle was also tied to the sycl queue.

Is it the worry that a given host_task thread will create `cublasHandle 0` tied to queue 0 or in another words: to device 0 but then that thread might be later reused for another queue, e.g. queue 1. In that case this thread might not be destroyed before this reusing and hence `cublasHandle 0` won't be destroyed. For that reason we have it destroyed when queue 0 is destroyed.

Reply:
I think this causes problems if that thread can in fact be reused for another device. Otherwise I guess it is fine since we would still preserve the mapping of `unique cublas handl -> unique device`

Reply:
The more I think about this the more issues I'm finding with the current version of this PR. 
You've mentioned that:
> A typical application will have a main thread which will submit multiple host_task. From there it is hard to predict how many threads will be used, it depends on the SYCL runtime and on the graph dependencies. I believe DPC++ will try to re-use the same thread for 2 consecutive host_task with a linear dependencies.

So just to confirm, the main reason for `static thread_local` cublasHandle, is that a thread will be *reused*? And at the point when it is being reused, the handle will still be alive and we won't have to create it again?